### PR TITLE
[#5461] Implement EIP-712

### LINKED ISSUE #5461: Implement EIP-712
### Description

*Type*: Feature

*Summary*:  Implement ethereum/EIPs#712  (eth_signTypedData). 

As EIP-712 is  finally merged and both MetaMask and Trust Wallet will support it soon, it would be great if status would implement  this too.

### PR DESCRIPTION
**Pull request by flexsurfer**
fixes #5461

### CODE REVIEW

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java [hunk 1]
```diff
@@ -743,6 +743,25 @@ public void run() {
...
+
+        Runnable r = new Runnable() {
+            @Override
+            public void run() {
+                String res = Statusgo.signTypedData(data, password);
+                callback.invoke(res);
+            }
+        };
+
+        StatusThreadPoolExecutor.getInstance().execute(r);
```
Comment (jeluard) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:762]: Could be simplified by using an [inline lambda](https://www.oracle.com/webfolder/technetwork/tutorials/obe/java/Lambda-QuickStart/index.html)

File: src/status_im/ui/screens/wallet/send/events.cljs [hunk 2]
```diff
@@ -95,7 +105,10 @@
        (merge
         {:db (cond-> (assoc-in db' [:wallet :send-transaction] {})
 
-               (not= method constants/web3-personal-sign)
+               (not (#{constants/web3-sign-typed-data
+                       constants/web3-sign-typed-data-v3
+                       constants/web3-personal-sign}
```
Comment (jeluard) [inline src/status_im/ui/screens/wallet/send/events.cljs]: Looks duplicated? Can we reuse an extracted fn?

File: src/status_im/ui/screens/wallet/send/events.cljs [hunk 3]
```diff
@@ -157,13 +170,16 @@
            (models.wallet/open-modal-wallet-for-transaction db' transaction (first params)))
 
          ;;SIGN MESSAGE
-         (= method constants/web3-personal-sign)
-         (let [[address data] (models.wallet/normalize-sign-message-params params)]
+         (or (= method constants/web3-personal-sign)
+             (#{constants/web3-sign-typed-data constants/web3-sign-typed-data-v3} method))
```
Comment (jeluard) [inline src/status_im/ui/screens/wallet/send/events.cljs]: Same duplication

**Review by jeluard — CHANGES_REQUESTED** [review]

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 1 above)
Comment (flexsurfer) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:762]: @jeluard I just used code from other methods we have, do you think we need to change how it works in other methods? maybe we could do refactoring in a separate issue?

File: src/status_im/ui/screens/wallet/send/events.cljs (hunk 2 above)
Comment (flexsurfer) [reply src/status_im/ui/screens/wallet/send/events.cljs]: what do you mean by duplication?

Comment (jeluard) [reply src/status_im/ui/screens/wallet/send/events.cljs]: Several time the same checks, could be a separate fn?

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 1 above)
Comment (jeluard) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:762]: Right, separate issue :D

**Comment by flexsurfer** [comment]
@jeluard done

**Review by jeluard — APPROVED** [review]

**Review by dmitryn — APPROVED** [review]

**Comment by asemiankevich** [comment]
@flexsurfer would you mind to provide a valid JSON for Typed Message to check this in Test Dapp? Or you can pre-populate it in the text box?  Thanks!
image

i have found this https://rstormsf.github.io/js-eth-personal-sign-examples/ but can't figure out how to get those JSONs to test with Status instead of MM

**Comment by flexsurfer** [comment]
updated dapp with default value

**Pull request merged**